### PR TITLE
Add support for calling AWS services as Task states

### PR DESCRIPTION
**Pull request by derek-pryor**
In November 2018 AWS added a new ability
> Step Functions now integrates with some AWS services. You can now directly call and pass parameters to the API of these integrated services from a task state in the Amazon States Language.

This pull request added support for calling the support AWS service using the following format
```py
Service.Function()
    parameters:
        Key1: Value1
        Key2: Value2
    # The normal task state modifiers
```

There is a JSON file that holds the definition for all of the supported AWS services, the available functions that can be called, and the required and optional argument for those functions. Heaviside checks to make sure that all required arguments are provided and that any provided argument is a valid argument name. Heaviside __does not__ check or validate the values of these arguments.

This update is considered a major version change due to restructuring how ARN construction is handled. Previously a `Lambda` or `Activity` could provide a partial ARN and the rest of the ARN would be created for the user. Now `Lambda` and `Activity` tasks can only provide the lambda or activity name and if the user wants to specify the full ARN they can use the `Arn` task. This simplifies the internals and does away with a callback function.

To keep the needed functionality of being able to modify the `Lambda` and `Activity` names I added a more generic `StateVisitor` patter that exposes the raw AST object to the library user, allowing them to modify the Task ARN, as well as any other part of any Task State.

### CODE REVIEW

File: heaviside/ast.py [hunk 1]
```diff
@@ -247,9 +288,70 @@ class ASTStateTask(ASTState):
...
+
+        if service.value == 'Arn' and not function.value.startswith('arn:aws:'):
+            function.raise_error("ARN must start with 'arn:aws:'")
+        if len(args_kwargs.args) > 0:
+            args_kwargs.args[0].raise_error('Unexpected argument')
+        if service.value in ('Lambda', 'Activity') and len(args_kwargs.kwargs) > 0:
+            tuple(args_kwargs.kwargs.keys())[0].raise_error('Unexpected keyword argument')
+
+        if service.value not in ('Lambda', 'Activity', 'Arn'):
+            required = AWS_SERVICES[service.value][function.lookup]['required_keys']
```
Comment (movestill) [inline heaviside/ast.py]: Do you need to make a copy of this since you're mutating it below?

Comment (derek-pryor) [reply heaviside/ast.py]: Good catch

File: docs/StepFunctionDSL.md [hunk 2]
```diff
@@ -173,40 +173,75 @@ Modifiers:
...
+
+AWS maintained APIs are in the format `_Type_._Function_()` and use the
+`parameters` block to pass argument.
+
+User maintained works are either `Lambda()` or `Activity()`, with the difference
+being where the code that will be executed is living. For `Lambda()` the code
+is an AWS Lambda function. For `Activity()` the code can be running anywhere,
+and is responsible for polling AWS to see if there is new work for it to perform.
+
+??? Do Lambda and Activity make use of the new parameters block?
```
[the remark this answers is not in the document]
Comment (derek-pryor) [reply docs/StepFunctionDSL.md]: No, it was/is a question, similar to TODO but different

From my reading of the documentation `parameters` are not used by `Activities` and `Lambdas`

File: heaviside/ast.py [hunk 3]
```diff
@@ -118,6 +133,25 @@ class ASTModOutput(ASTModKV):
 class ASTModData(ASTModKV):
     name = 'Data'
 
+class ASTModParameters(OrderedDict, ASTNode):
+    name = 'Parameters'
+
+    def __init__(self, parameters, kpv, kpvs):
```
Comment (movestill) [inline heaviside/ast.py]: Can you comment `kpv` and `kpvs`?

File: heaviside/ast.py [hunk 4]
```diff
@@ -574,3 +689,45 @@ def verify_goto_targets(branch):
...
+
+    def visit(self, branch):
+        """Visit all states in all branches of the state machine and dispatch
+        them to be handled the subclass
+
+        Args:
+            branch (list): List of ASTState objects
+        """
+        if not hasattr(branch, 'states'):
+            raise ValueError("Tryping to visit non-branch state: {}".format(branch))
```
Comment (movestill) [inline heaviside/ast.py]: Typo.

File: heaviside/ast.py (hunk 3 above)
Comment (derek-pryor) [reply heaviside/ast.py]: Done

**Pull request merged**